### PR TITLE
fixes #7332 - Host Create API documentation missing required parameters

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

**Comment by domcleal** [comment]
Many of these are incorrect, they're not required depending on various other parameters.  e.g. they might be supplied by a host group, or aren't required in !unattended mode, or aren't required if IP aren't required etc.

See also http://projects.theforeman.org/issues/6997, which I might close.

**Comment by isratrade** [comment]
@domcleal, so we should probably document the situations in :desc => when each param is required and when not. I agree with Vincent in the foreman-users thread. The docs are not user friendly as is.

**Comment by domcleal** [comment]
Alright then, let's see what happens.

**Comment by isratrade** [comment]
@domcleal, do you think it makes more since to have `:required => true` and in `:desc =>` explain when it's not true, or the opposite?

**Comment by domcleal** [comment]
The opposite, don't set required => true if it isn't always true.

**Comment by isratrade** [comment]
@domcleal, updated PR with new `:desc =>` when parameters are required.

**Comment by isratrade** [comment]
[test]

**Comment by isratrade** [comment]
@ares, can you review this as it's indirectly related to your PR #1766 about host validations

**Comment by isratrade** [comment]
@ares, ping. Can you please review this PR?

File: app/controllers/api/v2/hosts_controller.rb [hunk 1]
```diff
@@ -32,21 +32,22 @@ def show
...
-          param :environment_id, String
           param :location_id, :number, :required => true, :desc => N_("required if locations are enabled") if SETTINGS[:locations_enabled]
           param :organization_id, :number, :required => true, :desc => N_("required if organizations are enabled") if SETTINGS[:organizations_enabled]
+          param :environment_id, String, :desc => N_("required if host is managed and value is not inherited from hostgroup")
           param :ip, String, :desc => N_("not required if using a subnet with DHCP proxy")
-          param :mac, String, :desc => N_("not required if it's a virtual machine")
-          param :architecture_id, :number
+          param :mac, String, :desc => N_("required for managed host that is Bare Metal, not required if it's a virtual machine")
+          param :architecture_id, :number, :desc => N_("required if host is managed and value is not inherited from hostgroup")
           param :domain_id, :number
```
Comment (orrabin) [inline app/controllers/api/v2/hosts_controller.rb]: you need to add this `:desc => N_("required if host is managed and value is not inherited from hostgroup")` to here too.

``` ruby
 validates :architecture_id, :operatingsystem_id, :domain_id, :presence => true, :if => Proc.new {|host| host.managed}
```

File: app/controllers/api/v2/hosts_controller.rb [hunk 2]
```diff
@@ -32,21 +32,22 @@ def show
       def_param_group :host do
         param :host, Hash, :required => true, :action_aware => true do
           param :name, String, :required => true
-          param :environment_id, String
           param :location_id, :number, :required => true, :desc => N_("required if locations are enabled") if SETTINGS[:locations_enabled]
           param :organization_id, :number, :required => true, :desc => N_("required if organizations are enabled") if SETTINGS[:organizations_enabled]
+          param :environment_id, String, :desc => N_("required if host is managed and value is not inherited from hostgroup")
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: "host group"

(ditto in all strings)

File: app/controllers/api/v2/hosts_controller.rb [hunk 3]
```diff
@@ -32,21 +32,22 @@ def show
...
           param :ip, String, :desc => N_("not required if using a subnet with DHCP proxy")
-          param :mac, String, :desc => N_("not required if it's a virtual machine")
-          param :architecture_id, :number
+          param :mac, String, :desc => N_("required for managed host that is Bare Metal, not required if it's a virtual machine")
```
Comment (domcleal) [inline app/controllers/api/v2/hosts_controller.rb]: lower case "bare metal"

(ditto in all strings)

File: app/controllers/api/v1/hosts_controller.rb [hunk 4]
```diff
@@ -24,18 +24,22 @@ def show
...
         param :puppet_class_ids, Array
-        param :operatingsystem_id, String
+        param :operatingsystem_id, String, :desc => N_("required if host is managed and value is not inherited from hostgroup")
         param :medium_id, :number
-        param :ptable_id, :number
+        param :ptable_id, :number, :desc => N_("required if host is managed and custom partition has not been defined")
         param :subnet_id, :number
-        param :compute_resource_id, :number
+        param :compute_resource_id, :number, :desc => N_("nil means host is Bare Metal")
+        param :root_pass, String, :desc => N_("required if host is managed and value is not inherited from hostgroup or Settings default root password")
```
Comment (domcleal) [inline app/controllers/api/v1/hosts_controller.rb]: "or default password in settings"

File: app/controllers/api/v2/hosts_controller.rb [hunk 5]
```diff
@@ -32,21 +32,22 @@ def show
...
+          param :architecture_id, :number, :desc => N_("required if host is managed and value is not inherited from hostgroup")
           param :domain_id, :number
           param :realm_id, :number
           param :puppet_proxy_id, :number
           param :puppet_class_ids, Array
-          param :operatingsystem_id, String
+          param :operatingsystem_id, String, :desc => N_("required if host is managed and value is not inherited from hostgroup")
           param :medium_id, :number
```
Comment (orrabin) [inline app/controllers/api/v2/hosts_controller.rb]: missed this one:

``` ruby
validates :medium_id, :presence => true, :if => Proc.new { |host| host.validate_media? }
```

File: app/controllers/api/v1/hosts_controller.rb [hunk 6]
```diff
@@ -46,9 +50,10 @@ def show
         param :build, :bool
         param :enabled, :bool
         param :provision_method, String
-        param :managed, :bool
-        param :progress_report_id, String, :desc => 'UUID to track orchestration tasks status, GET /api/orchestration/:UUID/tasks'
+        param :managed, :bool, :desc => N_("this flag if a host is managed or unmanaged determines whether several parameters are required or not")
```
Comment (domcleal) [inline app/controllers/api/v1/hosts_controller.rb]: that's a side effect of the flag, that's not what the flag _means_

File: app/controllers/api/v1/hosts_controller.rb [hunk 7]
```diff
@@ -46,9 +50,10 @@ def show
...
+        param :progress_report_id, String, :desc => N_("UUID to track orchestration tasks status, GET /api/orchestration/:UUID/tasks")
```
Comment (domcleal) [inline app/controllers/api/v1/hosts_controller.rb]: v1 API isn't internationalised, we shouldn't extract strings or it causes more work for translators

**Comment by orrabin** [comment]
This tests fine but I think it is a bit confusing that we have `optional` under the param name and then a description on when it is required but that should probably be fixed in a separate PR

**Comment by isratrade** [comment]
@domcleal, re-committed per your comments.
@orrabin, I agree that it may be confusing to have `optional` under the param name and then a description on when it is required, but other option is to have it `required` and then write when it's optional.

**Comment by domcleal** [comment]
Setting required => true and then saying it's sometimes optional won't be compatible with apipie-bindings' validation of required parameters.

**Comment by dLobatog** [comment]
:+1: in my opinion, I'll fix the "Fase" -> "false" typos on merge. Probably this PR showcases the need for a better `require` syntax that is more than a simple boolean in apipie? cc @mbacovsky @iNecas @tstrachota

**Comment by dLobatog** [comment]
Merged as https://github.com/theforeman/foreman/commit/5638cc6871ed231c259ae64e998ff5907bfb7bb5 , thanks @isratrade !

**Pull request closed without merging**